### ISSUE TITLE
[GENERAL SUPPORT]: Seeking Guidance on Surrogate Model and Acquisition Function for High-Dimensional Optimization Problem

### ISSUE THREAD

**Issue by vinaysaini94** [facebook/Ax#2753]
### Question

Dear Research Community,

I am working on a project aimed at optimizing parameter values for a mechanical system to achieve the best possible outcomes. The system only allows one set of parameter values to be tested at a time, and we are dealing with more than 50 parameters, each within the range [1, 100].

Given the complexity of the problem, I have selected the SAASBO (Sparse Axis-Aligned Subspace Bayesian Optimization) model for optimization, with a total of 100 evaluations (n_init=10, n_batch=30, batch_size=3). Each evaluation takes approximately 30 minutes (~15 minutes for the surrogate model and ~15 minutes for the acquisition function). So overall time for all evaluations is roughly one complete day. However, sometimes the process does not yield optimal parameter values. 

I am seeking advice on the best surrogate model and acquisition function choices for this high-dimensional problem. Despite my initial expectation that the process would benefit from GPU acceleration, I have observed minimal GPU utilization during both the surrogate modeling and acquisition function stages, which has not led to significant speed improvements.

I would greatly appreciate any insights or suggestions on how to optimize this process, particularly in terms of model and acquisition function selection or any strategies to improve computational efficiency.

Thank you in advance for your support!

### Please provide any relevant code snippet if applicable.

```shell
model = Models.BOTORCH_MODULAR(
                 experiment=experiment,
                 data=data,surrogate=Surrogate(botorch_model_class=SaasFullyBayesianSingleTaskGP,
                                               mll_options={ "num_samples": 256,"warmup_steps": 512,},))
```

### Code of Conduct

- [X] I agree to follow this Ax's Code of Conduct

**Comment by danielcohenlive** [comment]
cc @Balandat

**Comment by hvarfner** [comment]
Hi @vinaysaini94,

It sounds like you have made sensible choices for your problem. However, if you could provide a slightly larger snippet, we could better ensure that the problem is set up correctly. For example, it could be the case that your input parameters are better represented on a log-scale, or that you have failed trials that produce extreme values and consequently compromise the model fit.

__Runtime:__ If you want to bring down the time of each BO iteration, I would recommend to substitute the SaasFullyBayesianSingleTaskGP for a SingleTaskGP and run the [dimension-scaled priors](https://github.com/pytorch/botorch/discussions/2451) that makes the SingleTaskGP performant on high-dimensional problems out of the box. This model does not use NUTS for sampling from the model posterior, and runs much faster as a result. _For now, this does require you install the current_ `main` _branch of BoTorch._

__Sample Efficiency:__ The SAAS model assumes that there is sparsity in your objective, so that only a few parameters meaningfully impact the outcome. If this is not the case for your problem, you could once again consider the dimension-scaled priors. If you do not think that your problem is sparse, this is potentially the better choice.

**Comment by vinaysaini94** [comment]
Hi @hvarfner ,
Thank you so much for your insightful suggestions and recommendations. I appreciate the detailed explanation, particularly regarding the potential benefit of using dimension-scaled priors with the SingleTaskGP model. I will certainly explore switching to SingleTaskGP to see how it impacts both runtime and performance, especially given the faster runtime without NUTS sampling.
 As requested, I’m sharing a larger code snippet below, along with some related files that are utilized in the setup. Hopefully, this provides more context, and you can suggest relevant modifications to improve the process.

Main code
"import atexit
import os
import csv
import time
import numpy as np
import torch
import pandas as pd
import json
from botorch.models.fully_bayesian import SaasFullyBayesianSingleTaskGP
from botorch.acquisition.logei import (
    qLogExpectedImprovement,
    qLogNoisyExpectedImprovement,
)
from ax.models.torch.botorch_modular.surrogate import Surrogate
from ax.modelbridge.random import RandomModelBridge
from ax import Experiment
from typing import Dict, cast
from ax.core.data import Data
from ax.core.arm import Arm
from ax.core.types import TTrialEvaluation
from ax.core.generator_run import GeneratorRun
from ax.core.objective import Objective
from ax.core.optimization_config import OptimizationConfig
from ax.modelbridge.registry import Models
from machine_folder.opt import OptSimulator, OptRunner
from machine_folder.metric import PlanQualityMetric
import machine_folder.opt_space as opt_space
from machine_folder.storage import DataStorage
import machine_folder.utils as utils


def main(n_batch=10, machine_id=pa,
         course_id="Test_Course", test_id="Saasbo",
         model_select='FULLYBAYESIAN',
         goal_fn="./config/PQMFinal.csv",
         obj_fn="./config/parameters_range.csv",
         const_fn = "./config/parameter_constraint.txt",
         n_init=10,
         batch_size=3,
         opt_dir=None,
         dvh_dir=None,
         **kwargs
         ):
    config_fn = kwargs.pop("env_fn", "./config/machine_config.json")
    with open(config_fn, "r") as buf:
         config = json.load(buf)
    tkwargs = {
        "dtype": torch.double,
        "device": torch.device("cuda" if torch.cuda.is_available() else "cpu")}
    if tkwargs["device"].type == "cuda":
            print(f"tkwargs is using CUDA.")
    else:
            print(f"tkwargs is not using CUDA.")
    goal_df = pd.read_csv(goal_fn)
    f_df, p_dict, f_dict = opt_space.parse_obj(obj_fn)
    
    c_str = opt_space.load_constraints(const_fn)
    struct_roi = {i.split("_")[0] for i in goal_df["Structure"]}
   
    app = api.CreateApplication("BayOpt")
    atexit.register(app.Dispose)
    machine = app.OpenmachineById(machine_id)
    machine.BeginModifications()
    opt_simulator = OptSimulator(plan, struct_roi, fixed_objectives=f_dict)
    opt_runner = OptRunner(opt_simulator)
    opt_prefix = "./opt_res/{}".format(machine_id) + opt_dir
    dvh_prefix = "./opt_res/{}".format(machine_id) + dvh_dir
    if not os.path.exists(opt_prefix):
        os.makedirs(opt_prefix)
    if not os.path.exists(dvh_prefix):
        os.makedirs(dvh_prefix)        
    save_data = DataStorage(opt_prefix, dvh_prefix, p_dict)  
    pars_random = []
    # initialize search space
    search_space = opt_space.form_space(p_dict, c_strs=c_str)

    plan_quality = PlanQualityMetric("PQM", goal_df, noise_sd=0)

    optimization_config = OptimizationConfig(objective=Objective(metric=plan_quality))
    
    print(f"doing {n_init + n_batch * batch_size} evaluations")
    experiment = Experiment(name="saasbo_planning",
                            search_space=search_space,
                            optimization_config=optimization_config,
                            runner=opt_runner)

    sobol = Models.SOBOL(search_space=experiment.search_space)
    
    experiment = opt_space.init_data(machine_id, experiment, n_init, mode="sobol", file='', generator=sobol) 
    
    for trial_name, trial in experiment.trials.items():
        pars_random.append(list(trial.arms[0].parameters.values())) 
        save_data.SAVE(trial.arms) 

    data = experiment.fetch_data()
 
    save_data.save_init_data(data, pars_random)  

    for i in range(n_batch):
        if model_select == 'FULLYBAYESIAN':
            model = Models.BOTORCH_MODULAR(
                 experiment=experiment,
                 data=data,
                 surrogate=Surrogate(botorch_model_class=SaasFullyBayesianSingleTaskGP,
                                               mll_options={ "num_samples": 256,"warmup_steps": 512,},),botorch_acqf_class=qLogExpectedImprovement)
        elif model_select == 'GPEI':
            model = Models.GPEI(
                experiment=experiment,
                data=data
            )
        elif model_select == 'SOBOL':
            model = Models.SOBOL(
                experiment=experiment,
                data=data
            )

        generator_run = model.gen(batch_size)

        trial = experiment.new_batch_trial(generator_run=generator_run)
        trial.run()
        trial.complete()
        data = Data.from_multiple_data([data, trial.fetch_data()])

        new_value = trial.fetch_data().df["mean"].max()
        if model_select == 'FULLYBAYESIAN':
            save_data.save_lenscal_saas(model)
        elif model_select == 'GPEI':
            save_data.save_lenscale_gpei(model)
        save_data.save_opt_data(data, trial, batch_size, n_init, i)
        save_data.SAVE(trial.arms)
        print(f"Iteration: {i}, Best in iteration {new_value:.3f}, Best so far: {data.df['mean'].max():.3f}")
    save_data.close_lenscale_file()

    best_params, arm_name = utils.extract_best_params(data, experiment)
    # last step: calculation and re-optimization with the best params
    data, trial = utils.calculate(experiment, arm_name, best_params, data)
    save_data.save_cal_data(data, best_params)
    save_data.SAVE(trial.arms, 'reopt-cal1')

    # reopt with calculated details
    data, trial = utils.reopt_calculated(experiment, best_params, data)
    save_data.save_cal_data(data, best_params)
    save_data.close_pars_file()
    save_data.SAVE(trial.arms, 'reopt-cal2')
    app.SaveModifications()
    app.Closemachine()
    app.Dispose()
    print("optimization finished")

"
I am also attaching the parameters range file, constraint file, Quality evaluation file.
Attached Files:
[Constraint.txt](https://github.com/user-attachments/files/16945941/constraint.txt)
[Parameters.csv](https://github.com/user-attachments/files/16945942/Paramters.csv)
[QMF.csv](https://github.com/user-attachments/files/16945943/QMF.csv)

Some parameters have a significant impact on the performance of the mechanical system, and its quality is evaluated based on the final parameters achieved using the values defined in the QMF files. The system's performance is assessed based on the objectives set in these files, and I hope that these additional details help in suggesting improvements.


Once again, I sincerely appreciate your guidance and support! I look forward to hearing your thoughts on the code snippet and further suggestions.

Best regards,
Vinay Saini

**Comment by hvarfner** [comment]
Hi @vinaysaini94,

From what I can tell, the setup looks okay. 

It sounds like your objective function is both custom-made and quite intricate, so I would give some extra consideration to whether it is reasonably continuous in your parameters. For example:

If your objective is to minimize the difference to the standard values (or if I'm guessing, the absolute value of the product of all of them), your objective value may change very rapidly in some parts of the search space, and barely at all in others. This characteristic is not well-handled by the GP, since it expects the objective function to be approximately as smooth throughout your entire search space. 

If this is the case for your objective, a simple log-transform should hopefully improve things.

**Comment by vinaysaini94** [comment]
I would like to add more details about the objective:
The objective function aim is to achieve maximum plan quality value by comparing what values is achieved by the mechanical system's, resultant-parameters with the value defined in the QMF.csv file. In which some of the parameters has intention to achieve higher value which has the priority 1, has intention o

**Comment by hvarfner** [comment]
Hi @vinaysaini94 

That seems like a very sensible definition. However, the same logic as above would apply to each component $f_i$ of $g$, so if these functions are non-smooth, or excessively steep, you would potentially lose efficiency in the optimization. 

Assert that the $f_i$'s are sensibly smooth, and apply transforms (such as `$\text{log}$`) if you think it improves things. With that said, I think you have defined the problem well based on what I have seen.

**Comment by vinaysaini94** [comment]
Hi @hvarfner,

Thank you for the information! I’d like some guidance on implementing the SingleTaskGP with dimension-scaled priors, as mentioned in the previous reply. Is it a matter of modifying a specific library module in the main code, or will I need to adjust certain lines of code as well? I’ve shared the relevant snippets in earlier comments for reference.

Once again, I sincerely appreciate your guidance and support! I look forward to hearing your thoughts on the code and any further suggestions.

Best regards,
Vinay

**Comment by hvarfner** [comment]
Sure! 

Option 1: Re-install [BoTorch](https://github.com/pytorch/botorch) from source, since it contains the necessary [changes to the hyperparameters priors](https://github.com/pytorch/botorch/discussions/2451) of the SingleTaskGP by default. Then, you're good to go.

Option 2:
Make the following manual edits to the code:

```
from math import log

from ax.modelbridge.registry import Models
from ax.models.torch.botorch_modular.surrogate import Surrogate
from botorch.models import SingleTaskGP
from gpytorch.constraints import GreaterThan
from gpytorch.kernels import RBFKernel
from gpytorch.priors import LogNormalPrior



SQRT2, SQRT3 = 2 ** 0.5, 3 ** 0.5


#################
PROBLEM_DIM = 50 # Please check that this is correct!
#################

lengthscale_prior = LogNormalPrior(SQRT2 + log(PROBLEM_DIM) * 0.5, SQRT3)
noise_prior = LogNormalPrior(-4.0, 1.0)
lengthscale_constraint = GreaterThan(2.5e-2, transform=None, initial_value=lengthscale_prior.mode)


model_step = Models.BOTORCH_MODULAR(
    experiment=experiment,
    data=data,
    surrogate=Surrogate(
        botorch_model_class=SingleTaskGP,
        covar_module_class=RBFKernel,
        covar_module_options={
            "lengthscale_prior": lengthscale_prior, 
            "lengthscale_constraint": lengthscale_constraint
        },
        likelihood_options={"noise_prior": noise_prior},
    )
)
```

**Comment by vinaysaini94** [comment]
Hi @hvarfner,

Thanks for the recent code modifications! They significantly improved both the speed and the overall quality of the results.

I just need a bit more clarification on your earlier suggestion to apply a log transformation to the objective values. Are you suggesting that I apply the log to each component of f_i, which ultimately forms g and then defines the metric (i.e., plan quality), which is used in the line optimization_config = OptimizationConfig(objective=Objective(metric=plan_quality))?

Or are you referring to the input parameters in the search space that are "range" type, floating, and continuous between [1.0, 100.0]? Specifically, should I modify this code line search_space = opt_space.form_space(p_dict, c_strs=c_str) to introduce the log transformation on each input parameter (generated using a custom file) in the range [1, 100] via SOBOL operators?

Once again, I sincerely appreciate your guidance and support! I look forward to hearing your thoughts and any further suggestions.

Best regards,
Vinay

**Comment by hvarfner** [comment]
Hi @vinaysaini94,

I'm glad to hear that the change improved things for you.

**Tranforming the inputs:** With the information I currently have, I unfortunately cannot say for sure since I don't know how your input parameters behave. _If_ you think a change from 1 to 10 is as meaningful as a change from 10-100, you should probably log the input parameters. You can do this by setting a parameter-wise `"log_scale": True`, example [here](https://github.com/facebook/Ax/blob/main/tutorials/tune_cnn_service.ipynb).

**Transforming the output values:** Same answer here, since I don't know what your $f_i$:s look like or what outputs are generally expected from your system. Please see my earlier answer on this topic to know what to look for with regard to transforming the outputs.

Best of luck!

**Comment by hvarfner** [comment]
@vinaysaini94 I hope the optimization works out well for you! If you don't have any additional questions, I'm going to go ahead and close this issue in a couple of days. Otherwise, ask away!

**Comment by vinaysaini94** [comment]
Dear @hvarfner 

Yes, this approach worked well for me! However, I need a bit more information. I applied the log scale to the range parameter values, and it worked fine as long as I didn’t use parameter constraints. When I attempted to use parameter constraints alongside the log-scaled parameters, I encountered an error message saying, *"Parameter constraints not allowed on log scale parameters."* 

It seems like I won’t be able to use both log scaling and constraints in my case.

Also, is there any module or method available for calculating the **Maximal Information Gain** for my problem? This information would be highly valuable for my work.

I appreciate any further insights you can provide on these issues!

Thank you again for your support.

Best regards,  
Vinay

**Comment by hvarfner** [comment]
@vinaysaini94 That is correct, constraints and log-space parameters are not supported.

To the second question, there is such method available. The MIG is not generally feasible to compute, since that requires you to know the smoothness and noise level of your objective. However, the hyperparameter prior is designed to assume your problem has a moderate MIG, so I would not worry to much about this.

**Comment by vinaysaini94** [comment]
Hi @hvarfner,

Thank you for the clarification regarding the log-space constraints and the Maximal Information Gain (MIG).

I have one more question: is there any theoretical basis or methodology for determining the optimal number of initial batches, the total number of batches, and the batch size for my optimization process? Alternatively, is there a module or method that can help calculate these values optimally?

Your guidance on this would be greatly appreciated!

Best regards,
Vinay

**Comment by hvarfner** [comment]
Hi @vinaysaini94,

In short: No, there isn't - but it's certainly an interesting area of research. Instead, I would recommend [early stopping](https://ax.dev/tutorials/early_stopping/early_stopping.html) if you are struggling with setting an appropriate budget for your experiment. Then, Ax will determine when optimization should stop based on the quality of the configurations that are found.

For batch size specifically, lower will always be better for sample efficiency, but worse for wall-clock time if you have the option to run multiple experiments in parallel. Thus, lower batch sizes (ideally 1) is the better choice if you are constrained on resources, and higher batch sizes are better if you want quicker results.
